File: java/Weed3/src/noear/weed/VariateEx.java
package noear.weed;

import noear.weed.ext.Act1;
import noear.weed.ext.Fun0;

/**
 * Created by yuety on 15/9/2.
 */
public class VariateEx extends Variate {
    protected Fun0<Object> _valueGetter;
    protected Act1<Object> _valueSetter;

    public VariateEx(String name,  Fun0<Object> valueGetter) {
        super(name, null);
        _valueGetter = valueGetter;
    }

    public VariateEx(String name, Fun0<Object> valueGetter, Act1<Object> valueSetter)  {
        super(name, null);
        this._valueGetter = valueGetter;
        this._valueSetter = valueSetter;
    }

    public  Object getValue() {
        return _value=_valueGetter.run();
    }

    public  DbType getType() {
        if (_value == null)
            _value = _valueGetter.run();

        return super.getType();
    }

    public  void setValue(Object value) {
        if (_valueSetter != null) {
            _valueSetter.run(value);
        }
    }
}

File: .net & mono/Weed3/VariateEx.cs
using System;
using System.Data;

namespace Noear.Weed {
    /**
 * Created by noear on 14-6-12.
 * 数据库访问参数（支付范型）
 */
    public class VariateEx : Variate {
        protected Func<Object>   _valueGetter;
        protected Action<Object> _valueSetter;

        public VariateEx(String name,  Func<Object> valueGetter) : base(name, null) {
            _valueGetter = valueGetter;
        }
        
        public VariateEx(String name, Func<Object> valueGetter, Action<Object> valueSetter) : base(name, null) {
            this._valueGetter = valueGetter;
            this._valueSetter = valueSetter;
        }

        public override Object getValue() {
            return _value=_valueGetter();
        }

        public override DbType getType() {
            if (_value == null)
                _value = _valueGetter();

            return base.getType();
        }

        public override void setValue(Object value) {
            if (_valueSetter != null) {
                _valueSetter(value);
            }
        }
    }
}
